Studio: Add new modal design

katinthehatsite opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

## Related issues

STU-226

## Proposed Changes

This PR adds a modal to Studio according to Figma: RToz6tIuQ7nlZrikBte4GU-fi-7169_205567

<img width="1255" alt="Screenshot 2025-03-26 at 9 57 33 AM" src="https://github.com/user-attachments/assets/aa1f44e7-a16d-426f-86e3-b186fd55064b" />

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Pull the changes from this branch
* Start the app with `STUDIO_WHATS_NEW_SECTION=true npm start`
* Click on the `Help > What's New` menu button in the topbar
* Observe that the modal appears
* Click through the modal and confirm that you can see intro page + three other pages with different features

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

Commit:
Use a new component for Modal

Commit:
Improve styling

Commit:
Final touches

Commit:
Handle on click

Commit:
Fix translations

Commit:
Update correct content

Commit:
Make code DRY

Commit:
Modify the button on finish

Commit:
Ensure the close button remains visible

katinthehatsite commented on src/components/main-sidebar.tsx:
```diff
@@ -33,9 +37,16 @@ export default function MainSidebar( { className }: MainSidebarProps ) {
+							className="text-white opacity-75 hover:opacity-100 text-sm w-full text-left px-2 py-1"
+						>
+							What's New
+						</button>
```
TODO: I added this for ease of testing, we will remove this before merging

ivan-ottinger replied:
That sounds good. 👌🏼 If we replace the temporary modal content (`src/modules/whats-new/components/whats-new-modal.tsx`) with your new one, we could also trigger the modal by removing `lastSeenVersion` from the `appdata-v1.json` and reloading the app.

katinthehatsite replied:
This is already done so I am going to mark this comment as resolved.

katinthehatsite commented:
@ivan-ottinger Feel free to give it an initial review while I am polishing a couple of things

ivan-ottinger commented on src/modules/whats-new-modal/whats-new-modal.tsx:
```diff
@@ -0,0 +1,76 @@
+			contentLabel={ __( "What's New in Studio" ) }
+			className="whats-new-modal !w-[312px] !h-[470px] overflow-hidden [&_.components-button_svg]:fill-white"
+			pages={ WHATS_NEW_PAGES.map( ( { image, ...pageContent } ) => ( {
+				image: <img src={ image } alt="" className="h-[173px] w-full object-cover mb-4" />,
```
Maybe we could also consider including the `title` text in the `alt` - if that makes sense from the accessibility point of view (e.g. when using VoiceOver).

I haven't tested this myself yet though.

katinthehatsite replied:
Added in https://github.com/Automattic/studio/pull/1128/commits/beaddc83442a159d9da3c59c4f4b35f713ea18fb

ivan-ottinger commented on src/modules/whats-new-modal/whats-new-modal.tsx:
```diff
@@ -0,0 +1,76 @@
+				image: <img src={ image } alt="" className="h-[173px] w-full object-cover mb-4" />,
+				content: <PageContent { ...pageContent } />,
+			} ) ) }
+			finishButtonText={ __( 'Done' ) }
```
I have noticed that the `Next` button is slightly taller than the `Done` button.

katinthehatsite replied:
Yeah that comes from the component itself, I am looking into fixing it 👀

katinthehatsite replied:
I fixed this in https://github.com/Automattic/studio/pull/1128/commits/a63ae6c5152c4db67f12c1c384f0f754c657ca56

I could not find a good way to fix in Tailwind so I had to use the last resort of CSS.

Commit:
Undo changes

ivan-ottinger reviewed (commented):
> @ivan-ottinger Feel free to give it an initial review while I am polishing a couple of things

Very nice! It great to see the modal coming to life! 🕺🏼 

Just sharing a couple smaller things I have noticed so far in the comments below.

I have also noticed the design has the `Previous` button in a border, but I think we are generally now trying to use borderless secondary buttons. We can double-check with Marina.

wojtekn commented:
> Click on What's New button in the sidebar

Shouldn't it be an item in the top menu?

Other feedback:
- the modal could use dots showing how many items there are, and which one is currently displayed
- I can't open the modal after I close it, and I need to restart Studio to do so.
- 'Learn more' and 'Previous' are displayed close to each other, so I could think they should be aligned, but they are not.
- Modal seems a bit too wide
- The feature doesn't obey the feature flag

ivan-ottinger commented:
> Shouldn't it be an item in the top menu?

Yes, there will be: https://github.com/Automattic/studio/pull/1129.

We have beeing using the sidebar temporarily in development. Kat had left a comment here yesterday: https://github.com/Automattic/studio/pull/1128/files#r2012860549

Also, I have still considered this PR as a work-in-progress, not ready for final review.

katinthehatsite commented:
> I have also noticed the design has the Previous button in a border, but I think we are generally now trying to use borderless secondary buttons. We can double-check with Marina.

Sure, we can consider adding it to work around the fact that Learn More and Previous might seem to close.

ivan-ottinger commented:
> I have also noticed the design has the Previous button in a border, but I think we are generally now trying to use borderless secondary buttons. We can double-check with Marina.

Here's where we were recently removing border from secondary buttons in Calypso: https://github.com/Automattic/wp-calypso/pull/101088. That was originally brought up in Staging sites walkthrough: pesfPa-1n1-p2. This is why I think it is the right approach to keep the secondary buttons without the border in Studio as well.

Commit:
Ensure that page controls are not affected

katinthehatsite commented:
> the modal could use dots showing how many items there are, and which one is currently displayed

The dots were actually there but I was fixing something else and accidentally changed their color. This was fixed in https://github.com/Automattic/studio/pull/1128/commits/713963864b98748f9b9633daf589ff03ff5d26b1

katinthehatsite commented:
> Here's where we were recently removing border from secondary buttons in Calypso: https://github.com/Automattic/wp-calypso/pull/101088. That was originally brought up in Staging sites walkthrough: pesfPa-1n1-p2. This is why I think it is the right approach to keep the secondary buttons without the border in Studio as well

Makes sense, should we get some feedback from Marina? Maybe we can add a light background or something? Not sure. I think the problem with no border is that it causes the alignment issue that @wojtekn mentioned in his feedback.

Commit:
Fix styling on modal

ivan-ottinger commented:
> Makes sense, should we get some feedback from Marina? Maybe we can add a light background or something? Not sure. I think the problem with no border is that it causes the alignment issue that @wojtekn mentioned in his feedback.

Yes, I will ask her in that Linear issue.

Commit:
Ensure that there is alt attribute on image

katinthehatsite commented:
> Modal seems a bit too wide 

Let's double-check with Marina, I followed what we have in Figma which is 312px. cc @ivan-ottinger

ivan-ottinger commented:
> Let's double-check with Marina, I followed what we have in Figma which is 312px. cc @ivan-ottinger

Maybe you meant too _tall_, @wojtekn? 🤔 Or do you mean that it would look better if it is a bit more narrow generally?

wojtekn commented:
My bad @ivan-ottinger , yes, I meant too narrow and too tall :- )

Commit:
Update link for custom domains

Commit:
Merge branch 'trunk' of github.com:Automattic/studio into add/whats-new-modal-new-component

Commit:
Use exisiting file structure

Commit:
Reuse logic for opening the modal

Commit:
Make sure that we have an initial intro card

katinthehatsite commented:
@ivan-ottinger @wojtekn I updated the testing steps and rebased the changes after https://github.com/Automattic/studio/pull/1129 was merged so that we can now manually trigger the modal (see above). Some remaining items:

* border / no border on the `Previous` button
* Not sure what is the copy on intro card
* Should the `Learn more` link on intro card link somewhere e.g. perhaps releases? or perhaps no Learn More button there?
* We are missing a couple of illustrations (marked in the comments in the code)
* The height / width of the modal (current dimensions use what we have in Figma, this might need to be double-checked further)

Otherwise, feel free to give it a review.

wojtekn commented:
> Not sure what is the copy on intro card

What about dropping the month from the heading and the instruction about clicking the card? Maybe something like this?

```
What is new in Studio?
Discover the latest updates in Studio! Explore exciting new features designed to enhance your experience.
```

> Should the Learn more link on intro card link somewhere e.g. perhaps releases? or perhaps no Learn More button there?

It think we can drop Learn more in the first screen as it's just intro explaining what other cards are.

I noticed two minor styling issues:
- What's New dialog gets a yellow outline, and other dialogs like Add site don't
- You can select text in the What's New dialog, e.g. using CMD+A, but we block it in other parts of the app

Let's Marina speak on dialog width and button border.

ivan-ottinger reviewed (commented):
Great work! The modal loads nicely and works as expected.

> border / no border on the Previous button

My apologies for that confusion around the button borders. I misunderstood the styling there initially. We would like to have the border in those buttons at the end. Related discussion: STU-270-linear-issue#comment-4bb5466b.

> Not sure what is the copy on intro card

I like the copy Wojtek suggested and we could still adjust it later.

> Should the Learn more link on intro card link somewhere e.g. perhaps releases? or perhaps no Learn More button there?

I agree here with Wojtek as well, that we can omit the button on the first intro item.

> We are missing a couple of illustrations (marked in the comments in the code)

When it comes to the illustration for the first intro card, I have asked Marina (STU-270-linear-issue#comment-8e9779f5).

For the illustration and copy we could use for the custom domains, here's a related discussion: p1742985723109549/1742552614.181789-slack-C06DRMD6VPZ. I think we should get to those soon - as the feature gets released.

> The height / width of the modal (current dimensions use what we have in Figma, this might need to be double-checked further)

→ Asked Marina in STU-270-linear-issue#comment-a943933d.

ivan-ottinger commented on src/modules/whats-new/components/whats-new-modal.tsx:
```diff
@@ -1,29 +1,96 @@
+import { __ } from '@wordpress/i18n';
+import { cx } from 'src/lib/cx';
+import { getIpcApi } from 'src/lib/get-ipc-api';
+import previewSitesIllustration from '../assets/preview-sites-illustration.svg';
+import versionSwitchIllustration from '../assets/version-switch-illustration.svg';
```
I think we could consider using absolute paths for imports:

```suggestion
import previewSitesIllustration from 'src/modules/whats-new/assets/preview-sites-illustration.svg';
import versionSwitchIllustration from 'src/modules/whats-new/assets/version-switch-illustration.svg';
```

katinthehatsite replied:
Updated in https://github.com/Automattic/studio/pull/1128/commits/84b352825fdb7a745f20ac76a97626ad02225096

Commit:
Use relative paths for import

Commit:
Add border to the Previous button

Commit:
Make learn more optional prop

Commit:
Adjust button styling

Commit:
Ensure modal text can't be select

Commit:
Remove focus outline

katinthehatsite commented:
I have now made the following updates:

* added border to `Previous` button
* removed focus and the ability to select text from the modal
* removed `Learn more` from the first card 

Remaining things:

* modal width / height clarification
* illustrations 
* copy details for domains and intro card 

cc @ivan-ottinger

ivan-ottinger commented:
Nice! Thank you for all the updates, Kat.

What if we used the following description for the custom domains card? We can still change it later if we decide to.

```
Easily identify your local Studio sites with custom domain names. Personalize and organize your workflow!
```

I think we could also use the copy suggested by Wojtek above for the first intro card:

```
What is new in Studio?
Discover the latest updates in Studio! Explore exciting new features designed to enhance your experience.
```

---

One small additional thing I noticed is that the `Learn more` link is a bit closer to the `Previous` button than in the Figma design (STU-270-linear-issue#comment-3ad4d247). I think we could add more space there so it stands out more.

Commit:
Update copy on domains

Commit:
Update first card copy

Commit:
Move the Learn more higher

Commit:
Ensure the Next button is dropped down

katinthehatsite commented:
I made some changes, feel free to take another look @ivan-ottinger . One thing that is tricky is that with translations, some content might be longer so I am slightly worried about scrollbars appearing. I am wondering if we might set max-height instead of fixed height

bgrgicak commented:
@katinthehatsite for long content you could make the footer sticky, so users could scroll the content, but still be able to navigate.

If you want to try it, you can do this in CSS:

```
.components-guide .components-modal__content {
    position: relative;
}

.components-guide__footer {
    position: sticky;
    bottom: 0;
    background: #fff;
    margin-bottom: 0;
    height: 104px;
    padding-top: 32px;
    display: flex;
    justify-content: space-between;
}
```


![Screenshot 2025-03-28 at 12 37 29](https://github.com/user-attachments/assets/53aaf87b-fcdb-4e10-906a-457300ce0190)
![Screenshot 2025-03-28 at 12 37 33](https://github.com/user-attachments/assets/4e415e2e-b9eb-4dd7-879f-da62500272f6)

katinthehatsite commented:
Thanks for the exploration, I am trying to avoid using CSS and looking for Tailwind approach. I do have the CSS solution already which would be something like:

```
.whats-new-modal .components-guide__container {
  display: flex;
  flex-direction: column;
  height: 100%;
}

/* Make the page container take full height */
.whats-new-modal .components-guide__page {
  display: flex;
  flex-direction: column;
  height: 100%;
}

/* Ensure content area can shrink if needed */
.whats-new-modal .components-guide__page-content {
  flex: 1;
  min-height: 0;
  display: flex;
  flex-direction: column;
}

/* Make the footer stick to the bottom */
.whats-new-modal .components-guide__footer {
  margin-top: auto;
}

/* Prevent text overflow by adding ellipsis */
.whats-new-modal h2,
.whats-new-modal p {
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  -webkit-line-clamp: 3; /* Limit to 3 lines */
  line-clamp: 3; /* Standard property for compatibility */
  -webkit-box-orient: vertical;
}

/* Allow title to show more */
.whats-new-modal h2 {
  -webkit-line-clamp: 2; /* Limit title to 2 lines */
  line-clamp: 2; /* Standard property for compatibility */
} 

.components-button.components-guide__finish-button {
	height: 40px;
}
```

I would also like avoid the scrollbars on the side. We should probably cut the text there.

Commit:
Use CSS to adjust modal

Commit:
Cleanup text

bgrgicak commented on src/modules/whats-new/components/whats-new-modal.tsx:
```diff
@@ -1,29 +1,100 @@
+		),
+		learnMoreUrl: 'https://wordpress.com/blog/2025/02/24/studio-preview-sites/',
+	},
+	{
+		image: previewSitesIllustration, // TODO: Add correct illustration
+		title: __( 'Choose a custom domain for your Studio site' ),
+		description: __(
+			'Easily identify your local Studio sites with custom domain names. Personalize and organize your workflow!'
+		),
+		learnMoreUrl:
+			'https://developer.wordpress.com/docs/developer-tools/studio/sites/#1-using-a-custom-domain ', // TODO: Add correct URL
+	},
```
There are some TODOs here, should we address them before merging?

katinthehatsite replied:
No these are fine, we are waiting for design. We will create a follow up issue to handle these

Commit:
Remove hard coded buttons

Commit:
Simplify CSS

ivan-ottinger commented:
Thank you for all those changes, Kat!

To mitigate the need for scrollbars, I wonder if we could make the modal height depending on the longest card content. 🤔 

Other approach coming to my mind is to make the modal a little bit larger in general to accommodate longer translations. The "What's New" modal in Calypso has this extra space that can be filled with the content without affecting the modal size or footer button positions:

![Markup on 2025-03-28 at 13:04:55](https://github.com/user-attachments/assets/657feca3-d9ba-468d-a7e4-39e7ef2626b4)

If the content gets larger than the buffer, in those cases it could just make the modal taller / or add a scrollbar.

---

In regards the use of CSS instead of Tailwind for this component in Studio, I wonder what is the point of view of other Yolo members.

---

One last comment: I think we could still merge the PR even if it is not 100% the way we want it since we are not yet launching to production. We could at least get the translations going and have the "What's New" modal ready for internal testing. It would also unblock the beta release. 🙂

Commit:
Fix clamping

katinthehatsite commented:
I agree that we can merge this as a base and then discuss the specifics. Currently, it will work like this:

* If the text is longer than 5 lines on the first card, it will be cut off to avoid using the scrollbar. This is as much as we can fit before we hit the footer. If it is shorter, it will not be cut off. The footer will remain at the bottom. 
* If the text is longer than 3 lines on all other cards, it will be cut off to avoid using scrollbars. The Learn More link will be displayed to see more info
* The modal is of the fixed height so it will not get scrollbars. Personally, I think it does not look great with those but I think it is a design call more than anything. 

I think a good approach from here would be to review this with the designer and see what improvements they could suggest to accommodate the translations etc. 

> In regards the use of CSS instead of Tailwind for this component in Studio, I wonder what is the point of view of other Yolo members.

I think in 95% of the time we should use Tailwind. However, this component is very tricky to handle as it allows minimal customizations. We can revisit if we can find an alternative approach for this.

ivan-ottinger commented:
Sounds good to me. 👍🏼 Thanks for sharing the details, Kat.

> I think a good approach from here would be to review this with the designer and see what improvements they could suggest to accommodate the translations etc.

Yes, let's discuss this with Marina. We can schedule a call with her for the beginning of the week when we get back from meetup - or - maybe we will find some time next week.
